projects/bms_carrier/test/test_passive_balance.c

```diff
@@ -0,0 +1,190 @@
+  .ltc_events = { .trigger_cell_conv_event = BMS_AFE_EVENT_TRIGGER_CELL_CONV,
+                  .cell_conv_complete_event = BMS_AFE_EVENT_CELL_CONV_COMPLETE,
+                  .trigger_aux_conv_event = BMS_AFE_EVENT_TRIGGER_AUX_CONV,
+                  .aux_conv_complete_event = BMS_AFE_EVENT_AUX_CONV_COMPLETE,
+                  .callback_run_event = BMS_AFE_EVENT_CALLBACK_RUN,
+                  .fault_event = BMS_AFE_EVENT_FAULT },
+  .cell_result_cb = NULL,
+  .aux_result_cb = NULL,
+  .result_context = NULL,
+  // literally random values
+  .mosi = TEST_LTC_AFE_SPI_MOSI,
+  .miso = TEST_LTC_AFE_SPI_MISO,
+  .sclk = TEST_LTC_AFE_SPI_SCLK,
+  .cs = TEST_LTC_AFE_SPI_CS,
+
+  .spi_port = TEST_LTC_AFE_SPI_PORT,
+  .spi_baudrate = TEST_LTC_AFE_SPI_BAUDRATE,
+  .adc_mode = TEST_LTC_AFE_ADC_MODE,
+};
+
+static LtcAfeStorage s_test_afe_storage;
+
+static const uint16_t AFE_DEFAULT_DISCHARGE_BITSET = 0;
+
+void setup_test(void) {
+  LOG_DEBUG("Setting up test sequence\n");
+  interrupt_init();
+  soft_timer_init();
+
+  // Configure all cells for output
+  for (uint8_t i = 0; i < LTC_AFE_MAX_DEVICES; i++) {
+    s_test_afe_settings.cell_bitset[i] = 1;
+    for (uint8_t j = 1; j < LTC_AFE_MAX_CELLS_PER_DEVICE; j++) {
+      s_test_afe_settings.cell_bitset[i] <<= 1;
+      s_test_afe_settings.cell_bitset[i] |= 0x1;
+    }
+  }
+
+  ltc_afe_init(&s_test_afe_storage, &s_test_afe_settings);
+  TEST_ASSERT_OK(passive_balance_init(&s_test_afe_storage));
+}
+
+void teardown_test(void) {}
+
+// Balance a few values, ensure cell bitset is set properly.
+void test_normal_operation(void) {
+  LOG_DEBUG("Testing passive balancing normal operation\n");
+  // Doing all tests on first 4 cells of device 0 or last cell for clarity, since max number of
+  // devices is variable
+  // Fill rest with numbers from 1010 to 1020
+  for (uint8_t i = NUM_TEST_VOLTAGES; i < LTC_AFE_MAX_CELLS; i++) {
+    s_test_afe_storage.cell_voltages[i] = 1010;
+  }
+
+  // Stores cell and device that should be balanced
+  uint8_t balance_cell = 2;
+  uint8_t balance_device = 0;
+
+  // For confirming test
+  uint16_t actual_cell = 0;
+  uint16_t device_cell = 0;
+
+  // Reset discharge bitset
+  s_test_afe_storage.discharge_bitset[balance_device] = AFE_DEFAULT_DISCHARGE_BITSET;
+  // Only cell 2 should be balanced
+  s_test_afe_storage.cell_voltages[0] = 1000;
+  s_test_afe_storage.cell_voltages[1] = 1025;
+  s_test_afe_storage.cell_voltages[2] = 1030;
+  s_test_afe_storage.cell_voltages[3] = 1010;
+
+  // Calculate device_cell in same way as ltc_afe_impl_toggle_cell_discharge would
+  // to check that bitset is as expected for the given cell
+  actual_cell = s_test_afe_storage.discharge_cell_lookup[balance_cell];
+  device_cell = actual_cell % LTC_AFE_MAX_CELLS_PER_DEVICE;
+  passive_balance(&s_test_afe_storage);
+  TEST_ASSERT_EQUAL(s_test_afe_storage.discharge_bitset[balance_device],
+                    AFE_DEFAULT_DISCHARGE_BITSET | (1 << device_cell));
```

Comment:
Also I feel like you shouldn't be looking at the internals of other modules like this ... might be better to just mock and verify it was called with the correct values. That or add a function for checking discharge bitset to the ltc_afe interface